src/sagemaker_xgboost_container/algorithm_mode/train.py

```diff
@@ -169,50 +152,82 @@ def sagemaker_train(
-        distributed.wait_hostname_resolution(sm_hosts)
+    is_dask_job = validated_train_config.pop("use_dask_gpu_training", "false")
 
-        if not train_dmatrix:
+    if is_dask_job == "true":
+        distributed_gpu_training.check_if_all_conditions_met(
+            tree_method_hp=tree_method_hp,
+            num_hosts=num_hosts,
+            num_gpus=num_gpus,
+            input_mode=input_mode,
+            input_format=file_type,
+            data_config=validated_data_config,
+        )
+        logging.info("Going to run distributed GPU training through Dask.")
+        distributed_gpu_training.run_training_with_dask(
+            hyperparameters=validated_train_config,
+            train_path=train_path,
+            validation_path=val_path,
+            model_dir=model_dir,
+            content_type=file_type,
+            sm_hosts=sm_hosts,
+            current_host=sm_current_host,
+            checkpoint_dir=checkpoint_dir,
+            num_gpus=num_gpus,
+        )
+    else:
+        if num_gpus > 1:
             logging.warning(
-                "Host {} does not have data. Will broadcast to cluster and will not be used in distributed"
-                " training.".format(sm_current_host)
+                f"If you're using GPU training, not all GPUs on the instance will be used. "
+                f"See how to use all GPUs at {DOCUMENTATION_LINK}"
             )
-        distributed.rabit_run(
-            exec_fun=train_job,
-            args=train_args,
-            include_in_training=(train_dmatrix is not None),
-            hosts=sm_hosts,
-            current_host=sm_current_host,
-            update_rabit_args=True,
+
+        train_dmatrix, val_dmatrix, train_val_dmatrix = get_validated_dmatrices(
+            train_path, val_path, file_type, csv_weights, is_pipe, combine_train_val
         )
-    elif num_hosts == 1:
-        if train_dmatrix:
-            if validation_channel:
-                if not val_dmatrix:
-                    raise exc.UserError("No data in validation channel path {}".format(val_path))
-            logging.info("Single node training.")
-            train_args.update({"is_master": True})
-            train_job(**train_args)
+        train_args = dict(
+            train_cfg=validated_train_config,
+            train_dmatrix=train_dmatrix,
+            val_dmatrix=val_dmatrix,
+            train_val_dmatrix=train_val_dmatrix,
+            model_dir=model_dir,
+            checkpoint_dir=checkpoint_dir,
+        )
+        if num_hosts > 1:
+            # Wait for hosts to find each other
+            logging.info("Distributed node training with {} hosts: {}".format(num_hosts, sm_hosts))
+            distributed.wait_hostname_resolution(sm_hosts)
+            if not train_dmatrix:
+                logging.warning(
+                    "Host {} does not have data. Will broadcast to cluster and will not be used in distributed"
+                    " training.".format(sm_current_host)
```

Comment:
For these files, feel free to install and run `pyupgrade --py38-plus`

Reply:
I'll run for the next PR to keep this one from having more changes.